client/components/gravatar/index.jsx

```diff
@@ -46,22 +37,29 @@ export class Gravatar extends Component {
 		size: 32,
 	};
 
+	state = { failedToLoad: false };
+
 	getResizedImageURL( imageURL ) {
 		const { imgSize } = this.props;
 		imageURL = imageURL || 'https://www.gravatar.com/avatar/0';
-		const parsedURL = url.parse( imageURL );
-		const query = parse( parsedURL.query );
+
+		let parsedURL;
+		try {
+			parsedURL = new URL( imageURL );
+		} catch {
+			// Invalid URL; return unmodified.
+			return imageURL;
+		}
```

Comment:
Particularly since we'll need to do this everywhere we use `URL` can we create a wrapper function that handles the exception? I'm not sure if there's anywhere we would need to differentiate an exception from a falsey value, and I'm predicting that people will call `new URL()` without the try/catch.

```js
const parseURL = s => {
	try {
		return new URL( s );
	} catch ( e ) {
		return null;
	}
}
```

```js
const parsedURL = parseURL( imageURL );
if ( ! parsedURL ) {
	return imageURL;
}
```

Reply:
I'm afraid it's not as simple as that. The URL parsing takes a second, optional parameter, for a base URL. This is because the `URL` object only supports absolute URLs, and so it needs a base URL to complete any partial URL that may be provided.

As such, we may want different behaviours, depending on whether or not that 2nd parameter is provided.

In this case, it's not, which means we assume that the provided URL will be an absolute URL and we want it to fail if not.

Reply:
In fact, let me explain things a bit deeper, since `URL` isn't trivial.

`URL` only handles absolute URLs. It's not able to represent a partial URL at all, even if just the protocol is missing (making it a scheme-relative URL). There is a second parameter in the `URL` constructor, which allows you to provide a base URL, which is used to resolve the missing bits in the first parameter. That is, if you have a path-absolute URL, for example, and provide a base URL, it'll take the protocol and domain from the base URL, while keeping the path from the first parameter:

```js
const url = new URL( '/my/path', 'https://wordpress.com/other/path');
url.href === 'https://wordpress.com/my/path';
```

However, the deprecated `parse` and `format` from `node`'s `url` are one of those magic APIs that handle all types of URLs without complaining (or without erroring out when they should, in many cases), which means that we have a lot of code that expects to be able to parse and format partial URLs of various sorts.

As such, I added two functions to `lib/url`: `determineUrlType` and `format`. `determineUrlType` determines the URL type for a provided URL: absolute (`http://wordpress.com/path`), scheme-relative (`//wordpress.com/path`), path-absolute (`/path`), path-relative (`path` or `../path`), or invalid (any invalid URL). `format` normalises a URL, optionally downcasts it to a lesser type (e.g. from absolute to scheme-relative) and outputs it as a string.

With all this in mind, I could definitely take the approach of reimplementing something similar to `node` `url`'s `parse` and `format` based on `URL`. It would be the easiest way to migrate code, with as few changes as possible. However, it seems to me that it could perpetuate some of mistakes of the `node` API:
- It's hard to tell what types of URL a piece of code that makes use of the API is expecting.
- It's hard to limit the types of URL you want to handle.
- It's easy to make mistakes, in thinking you're only handling absolute URLs, when you're actually handling everything.

I'm happy to discuss different ways of handling things more easily, without as much trying and catching going on. The other methods in `lib/url` (such as `addQueryArgs`, `omitURLParams`, etc.) already provide some help, by offering easier, focused high-level APIs, but I haven't come up with a good API for the general use-case that doesn't suffer from the same issues in `node`'s `url`.